Stop using -_contextMenuInteraction:styleForMenuWithConfiguration: in WKFileUploadPanel

whsieh opened the pull request:
#### 5d29c2a7ed7792a2af1fdd9270eb9c86aa89d725
<pre>
Stop using -_contextMenuInteraction:styleForMenuWithConfiguration: in WKFileUploadPanel
<a href="https://bugs.webkit.org/show_bug.cgi?id=262354">https://bugs.webkit.org/show_bug.cgi?id=262354</a>

Reviewed by Richard Robinson.

Introduce the `CompactContextMenuPresenter` helper class, which is used to present or dismiss a
UIContextMenu using compact layout; additionally, deploy this helper class in `WKFileUploadPanel` to
eliminate uses of both `-_contextMenuInteraction:styleForMenuWithConfiguration:` and
`_UIContextMenuLayoutCompactMenu`. See below for more details.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/CompactContextMenuPresenter.h: Added.
* Source/WebKit/UIProcess/ios/CompactContextMenuPresenter.mm: Added.

Implement a new helper class that encapsulates a hidden and non-interactable `UIButton`, as well as
its corresponding `UIContextMenuInteraction`. This helper class is constructed with a context menu
interaction delegate, to which the helper class forwards delegate method calls; this allows the
delegate to influence the hidden button&apos;s compact context menu as if it were its own, but defers the
compact layout to the button itself.

This allows a `CompactContextMenuPresenter` to (functionally) act as a drop-in replacement for an
actual `UIContextMenuInteraction`.

(-[WKCompactContextMenuPresenterButton contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:configuration:highlightPreviewForItemWithIdentifier:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:willDisplayMenuForConfiguration:animator:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:willEndForConfiguration:animator:]):

Forward delegate calls to the given external context menu delegate if applicable (see
`WKFileUploadPanel` below).

(WebKit::CompactContextMenuPresenter::CompactContextMenuPresenter):
(WebKit::CompactContextMenuPresenter::~CompactContextMenuPresenter):
(WebKit::CompactContextMenuPresenter::present):
(WebKit::CompactContextMenuPresenter::interaction const):
(WebKit::CompactContextMenuPresenter::dismiss):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(-[WKFileUploadPanel dealloc]):

Replace the existing `UIContextMenuInteraction`, `_documentContextMenuInteraction`, with a
`CompactContextMenuPresenter` instead.

* Source/WebKit/UIProcess/mac/WKTextFinderClient.mm:

Unrelated unified source build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268655@main">https://commits.webkit.org/268655@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df279382a90bda4a7b010195679d7ee295fef573

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20220 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20653 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22119 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18870 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20817 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20316 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20344 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22970 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17512 "8 flakes 3 failures") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24642 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18558 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22613 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19138 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16248 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18343 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4880 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22685 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18963 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/ios/CompactContextMenuPresenter.h:
```diff
@@ -0,0 +1,56 @@
+    UIContextMenuInteraction *interaction() const;
+
+private:
+    __weak UIView *m_rootView { nil };
```
is there a reason to prefer this over using a `WeakObjCPtr`?

whsieh replied:
It's generally cleaner than `WeakObjCPtr` (the latter was only necessary in WebKit2 before `OBJC_WEAK` was [enabled a few years ago](https://commits.webkit.org/200302@main))

whsieh committed:
Stop using -_contextMenuInteraction:styleForMenuWithConfiguration: in WKFileUploadPanel
https://bugs.webkit.org/show_bug.cgi?id=262354

Reviewed by Richard Robinson.

Introduce the `CompactContextMenuPresenter` helper class, which is used to present or dismiss a
UIContextMenu using compact layout; additionally, deploy this helper class in `WKFileUploadPanel` to
eliminate uses of both `-_contextMenuInteraction:styleForMenuWithConfiguration:` and
`_UIContextMenuLayoutCompactMenu`. See below for more details.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/CompactContextMenuPresenter.h: Added.
* Source/WebKit/UIProcess/ios/CompactContextMenuPresenter.mm: Added.

Implement a new helper class that encapsulates a hidden and non-interactable `UIButton`, as well as
its corresponding `UIContextMenuInteraction`. This helper class is constructed with a context menu
interaction delegate, to which the helper class forwards delegate method calls; this allows the
delegate to influence the hidden button's compact context menu as if it were its own, but defers the
compact layout to the button itself.

This allows a `CompactContextMenuPresenter` to (functionally) act as a drop-in replacement for an
actual `UIContextMenuInteraction`.

(-[WKCompactContextMenuPresenterButton contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:configuration:highlightPreviewForItemWithIdentifier:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:willDisplayMenuForConfiguration:animator:]):
(-[WKCompactContextMenuPresenterButton contextMenuInteraction:willEndForConfiguration:animator:]):

Forward delegate calls to the given external context menu delegate if applicable (see
`WKFileUploadPanel` below).

(WebKit::CompactContextMenuPresenter::CompactContextMenuPresenter):
(WebKit::CompactContextMenuPresenter::~CompactContextMenuPresenter):
(WebKit::CompactContextMenuPresenter::present):
(WebKit::CompactContextMenuPresenter::interaction const):
(WebKit::CompactContextMenuPresenter::dismiss):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(-[WKFileUploadPanel dealloc]):

Replace the existing `UIContextMenuInteraction`, `_documentContextMenuInteraction`, with a
`CompactContextMenuPresenter` instead.

* Source/WebKit/UIProcess/mac/WKTextFinderClient.mm:

Unrelated unified source build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/268655@main
